### PR TITLE
nce: add log_norm_factor for nce

### PR DESCRIPTION
**Pull request by Gerstenberger**
I propose the following changes to the `sampling_loss`:
* Add `log_norm_factor` to `sampling_loss` for nce; this is fixed to 0.0 in the standard TensorFlow implementation. However, we found that this is worth tuning for Noise Contrastive Estimation.
* As result, it is required to also include `compute_sampled_logits` and `sum_rows` separately, as this is an implementation detail in TensorFlow and not part of the public API. These functions are mostly copied from there. See comments in the doc-string.

### CODE REVIEW

File: TFNetworkLayer.py [hunk 1]
```diff
@@ -8257,6 +8258,7 @@ def __init__(self,
       and the objective is equal to sampled logistic loss.
     :param dict[str] sampler_args: additional arguments for the candidate sampler. This is most relevant to the fixed_unigram sampler.
       See https://www.tensorflow.org/api_docs/python/tf/random/fixed_unigram_candidate_sampler for details.
+    :param float log_norm_factor=0.0: The logarithm of the constant normalization term for NCE.
```
Comment (albertz) [inline TFNetworkLayer.py]: I think the format syntax is wrong (remove the `=0.0` here).

File: TFNetworkLayer.py [hunk 2]
```diff
@@ -8273,6 +8275,68 @@ def __init__(self,
...
+      https://github.com/tensorflow/tensorflow/blob/e19c354920c3b246dda6598229210a582caaa1a9/tensorflow/python/ops/nn_impl.py#L1729-L1837
+
+    :param tf.Tensor weights: NCE parameters of shape [num_classes, dim].
+    :param tf.Tensor biases: biases with shape [num_classes]
+    :param tf.Tensor labels: The target classes of shape [batch_size, num_true]
+    :param tf.Tensor inputs: The forward activations of the network of shape [batch_size, dim]
+    :param int num_sampled: Number of sampled classes
+    :param int num_classes: Number of possible classes
+    :param int num_true: Number of target classes per training example
+    :param (tf.Tensor[int64], tf.Tensor, tf.Tensor) sampled_values: a tuple of
```
Comment (albertz) [inline TFNetworkLayer.py]: I think `tf.Tensor[int64]` is invalid syntax. It should just be `tf.Tensor`. Write in the description that `sampled_candidates` is of dtype int64.

File: TFNetworkLayer.py [hunk 3]
```diff
@@ -8273,6 +8275,68 @@ def __init__(self,
...
+    :param int num_true: Number of target classes per training example
+    :param (tf.Tensor[int64], tf.Tensor, tf.Tensor) sampled_values: a tuple of 
+      (`sampled_candidates`, `true_expected_count`, `sampled_expected_count`)
+      returned by a `*_candidate_sampler` function.
+      (if None, we default to `log_uniform_candidate_sampler`)
+    :param bool remove_accidental_hits: Remove sampled classes that equal on of the target classes
+    :param str partition_strategy: A string specifying the partitioning strategy, relevant
+      if `len(weights) > 1`. Currently `"div"` and `"mod"` are supported.
+      Default is `"mod"`. See `tf.nn.embedding_lookup` for more details.
+    :param str name: Name for the operation
```
Comment (albertz) [inline TFNetworkLayer.py]: Should be `str|None`.

File: TFNetworkLayer.py [hunk 4]
```diff
@@ -8273,6 +8275,68 @@ def __init__(self,
...
+    :param int seed: Seed for sampling if no sampled_values are given
```
Comment (albertz) [inline TFNetworkLayer.py]: Should be `int|None`.

File: TFUtil.py [hunk 5]
```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
...
+
+
+def sum_rows(x):
+  """
+  Returns a vector summing up each row of the matrix x.
+
+  This is mostly a copy of
+    https://github.com/tensorflow/tensorflow/blob/e19c354920c3b246dda6598229210a582caaa1a9/tensorflow/python/ops/nn_impl.py#L1427
+
+  :param tf.Tensor x: input tensor
```
Comment (albertz) [inline TFUtil.py]: What shape do you expect for `x`? What dtype? Please specify.

File: TFUtil.py [hunk 6]
```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
...
+  :return: Vector, which rows are summed along axis 1 with shape [shape(x)[0]]
+  :rtype: tf.Tensor
+  """
+  # sum_rows(x) is equivalent to math_ops.reduce_sum(x, 1) when x is
+  # a matrix.  The gradient of _sum_rows(x) is more efficient than
+  # reduce_sum(x, 1)'s gradient in today's implementation. Therefore,
+  # we use _sum_rows(x) in the nce_loss() computation since the loss
+  # is mostly used for training.
+  cols = tf.shape(x)[1]
+  ones_shape = tf.stack([cols, 1])
```
Comment (albertz) [inline TFUtil.py]: You don't need the `tf.stack` here, or do you?

File: TFUtil.py [hunk 7]
```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
...
+
+  :param tf.Tensor x: input tensor
+  :return: Vector, which rows are summed along axis 1 with shape [shape(x)[0]]
+  :rtype: tf.Tensor
+  """
+  # sum_rows(x) is equivalent to math_ops.reduce_sum(x, 1) when x is
+  # a matrix.  The gradient of _sum_rows(x) is more efficient than
+  # reduce_sum(x, 1)'s gradient in today's implementation. Therefore,
+  # we use _sum_rows(x) in the nce_loss() computation since the loss
+  # is mostly used for training.
```
Comment (albertz) [inline TFUtil.py]: `sum_rows` or `_sum_rows`? Be consistent.
Also, maybe this description would actually be also helpful in the docstring itself, not just as a code comment? So move it there.

Comment (albertz) [reply TFUtil.py]: Btw, are you sure that the gradient is more efficient? Have you measured that? Or do they write that in the TF comment?
I remember seeing people using `matmul` instead of `reduce_sum`, but the reason was that in earlier TF versions, `matmul` was deterministic and `reduce_sum` was not (in later TF versions, also `reduce_sum` should be deterministic).

File: TFUtil.py [hunk 8]
```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
...
+  e.g. noise-contrastive estimation (see nce_loss) or sampled softmax (see
+  sampled_softmax_loss).
+  Note: In the case where num_true > 1, we assign to each target class
+  the target probability 1 / num_true so that the target probabilities
+  sum to 1 per-example.
+
+  This is mostly a copy of
+    https://github.com/tensorflow/tensorflow/blob/e19c354920c3b246dda6598229210a582caaa1a9/tensorflow/python/ops/nn_impl.py#L1440
+
+  :param weights tf.Tensor|[tf.Tensor]: A `Tensor` of shape `[num_classes, dim]`, or a list of `Tensor`
```
Comment (albertz) [inline TFUtil.py]: The syntax is wrong. It should be `:param tf.Tensor|list[tf.Tensor| weights: ...`.

File: TFUtil.py [hunk 9]
```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
...
+  :return:
+    out_logits: `Tensor` object with shape
+        `[batch_size, num_true + num_sampled]`, for passing to either
+        `nn.sigmoid_cross_entropy_with_logits` (NCE) or
+        `nn.softmax_cross_entropy_with_logits` (sampled softmax).
+    out_labels: A Tensor object with the same shape as `out_logits`.
+  :rtype: (tf.Tensor, tf.Tensor)
+  """
+
+  if not isinstance(weights, list):
```
Comment (albertz) [inline TFUtil.py]: Also allow `tuple`. (i.e. `isinstance(weights, (list, tuple))`).

File: TFUtil.py [hunk 10]
```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
...
+          validate_indices=False)
+
+    if subtract_log_q:
+      true_logits -= tf.log(true_expected_count)
+      sampled_logits -= tf.log(sampled_expected_count)
+
+    out_logits = tf.concat([true_logits, sampled_logits], 1)
+
+    out_labels = tf.concat([
+        tf.ones_like(true_logits) / num_true,
```
Comment (albertz) [inline TFUtil.py]: Use `//` for integer division.

File: TFUtil.py (hunk 7 above)
Comment (albertz) [reply TFUtil.py]: I just checked; that comment is copied from the TF code. So they claim that. But that code is at least 4 years old, probably even older, and dates back to before TF was public (see [here](https://github.com/tensorflow/tensorflow/blame/ef2a926ec05dfd337d84279aafa58b22f0f36123/tensorflow/python/ops/nn.py#L1084)). So I'm not sure if that is still true today. I would verify that if you have time (just try out with `tf.reduce_sum` instead, and see if there is a difference). In any case, replace "today", and be more specific (write the date of the initial TF commit).

File: TFUtil.py [hunk 11]
```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
...
+                           seed=None):
+  """Helper function for nce_loss and sampled_softmax_loss functions.
+  Computes sampled output training logits and labels suitable for implementing
+  e.g. noise-contrastive estimation (see nce_loss) or sampled softmax (see
+  sampled_softmax_loss).
+  Note: In the case where num_true > 1, we assign to each target class
+  the target probability 1 / num_true so that the target probabilities
+  sum to 1 per-example.
+
+  This is mostly a copy of
```
Comment (kazuki-irie) [inline TFUtil.py]: > This is **mostly** a copy of

What are the changes?

File: TFUtil.py (hunk 7 above)
Comment (Gerstenberger) [reply TFUtil.py]: I see, thanks.
I removed sum_rows now and replaced this with reduce_sum. I did some small scale tests and on the V100 there is no difference in runtime.

File: TFUtil.py (hunk 10 above)
Comment (Gerstenberger) [reply TFUtil.py]: I think this should be correct as is. If num_true > 1 and we do `floordiv` then the results should be `0.0` which is not what we want. This is just to ensure that the per-example labels sum to `1.0`

**Comment by Gerstenberger** [comment]
I did make the requested changes. Please let me know if anything is missing.

**Comment by albertz** [comment]
Btw, please clean up the commits in the end.

File: TFUtil.py [hunk 12]
```diff
@@ -8517,3 +8517,155 @@ def get_non_deterministic_ops_from_graph():
...
+  e.g. noise-contrastive estimation (see nce_loss) or sampled softmax (see
+  sampled_softmax_loss).
+  Note: In the case where num_true > 1, we assign to each target class
+  the target probability 1 / num_true so that the target probabilities
+  sum to 1 per-example.
+
+  This is a copy of
+    https://github.com/tensorflow/tensorflow/blob/e19c354920c3b246dda6598229210a582caaa1a9/tensorflow/python/ops/nn_impl.py#L1440
+
+  :param tf.Tensor|list(tf.Tensor)|tuple(tf.Tensor) weights: A `Tensor` of shape `[num_classes, dim]`,
```
Comment (albertz) [inline TFUtil.py]: The syntax is wrong. It should be `list[tf.Tensor]` and `tuple[tf.Tensor]`. (You can just check whether PyCharm correctly infers the type for `weights`.)

File: TFUtil.py [hunk 13]
```diff
@@ -8517,3 +8517,155 @@ def get_non_deterministic_ops_from_graph():
...
+    all_ids = tf.concat([labels_flat, sampled], 0)
+
+    all_w = tf.nn.embedding_lookup(
+        weights, all_ids, partition_strategy=partition_strategy)
+    if all_w.dtype != inputs.dtype:
+      all_w = tf.cast(all_w, inputs.dtype)
+
+    true_w = tf.slice(all_w,
+                      [0, 0],
+                      tf.stack([tf.shape(labels_flat)[0], -1]))
```
Comment (albertz) [inline TFUtil.py]: I think you don't need the `tf.stack` here.

File: TFUtil.py [hunk 14]
```diff
@@ -8517,3 +8517,155 @@ def get_non_deterministic_ops_from_graph():
...
+
+    sampled_w = tf.slice(
+        all_w, tf.stack([tf.shape(labels_flat)[0], 0]), [-1, -1])
```
Comment (albertz) [inline TFUtil.py]: I think you don't need the `tf.stack` here.

File: TFUtil.py (hunk 10 above)
Comment (albertz) [reply TFUtil.py]: Ah, sorry, I got confused. I thought that are labels (because you call it `out_labels`), i.e. integers. But why do you call it "labels" then? Better rename it, so that it is more clear what it is.

File: TFUtil.py [hunk 15]
```diff
@@ -8517,3 +8517,155 @@ def get_non_deterministic_ops_from_graph():
...
+    Default is `"mod"`. See `tf.nn.embedding_lookup` for more details.
+  :param str|None name: A name for the operation.
+  :param int|None seed: random seed for candidate sampling. Default to None, which doesn't set
+    the op-level random seed for candidate sampling.
+  :return:
+    out_logits: `Tensor` object with shape
+        `[batch_size, num_true + num_sampled]`, for passing to either
+        `nn.sigmoid_cross_entropy_with_logits` (NCE) or
+        `nn.softmax_cross_entropy_with_logits` (sampled softmax).
+    out_labels: A Tensor object with the same shape as `out_logits`.
```
Comment (albertz) [inline TFUtil.py]: What is in `out_labels`? Make more clear. Also write the dtype (from "labels", I would infer that it must be integer).

File: TFUtil.py (hunk 10 above)
Comment (albertz) [reply TFUtil.py]: I'm not sure if "targets" is better... Maybe "soft targets"? Well, maybe good enough like that.
